src/utils/StateUpdaters.tsx

```diff
@@ -1,101 +1,201 @@
     case "setSearchBarContent":
       return { ...state, searchBarContent: action.value as string };
     case "toggleVisibility":
       return { ...state, useMarkerVisibility: action.value as boolean };
     case "toggleSearchBarQuery":
       return { ...state, useSearchBar: action.value as boolean };
-    case "setInFullMode":
-      return { ...state, inFullMode: action.value as boolean };
+    case "setCurrentRow":
+      return { ...state, currentRow: action.value as Types.Row };
     default:
       return state;
   }
 };
 
-export const getUpdatedCombinedRowsByZoom = (
-  rows: Types.row[],
-  combinedRows: Types.combinedRow[],
+export const getJsonMarkers = (
+  dispatch: Types.Dispatch
+): React.EffectCallback => {
+  return () => {
+    const fetchMarkers = async (): Promise<void> => {
+      const response: Response = await fetch("/static/waypoints.json");
+      const markers: Types.Row[] = await response.json();
+      dispatch({ type: "setRows", value: markers });
+    };
+    fetchMarkers();
+  };
+};
+
+export const updateAllAndCurrentCombinedRows = (
+  stateManager: Types.StateManager
+): React.EffectCallback => {
+  const [state, dispatch] = stateManager;
+  return () => {
+    const updatedCombinedRows = getUpdatedCombinedRowsByZoom(
+      state.rows,
+      Config.defaultCombinedRows,
+      state.mousePosition.zoom
+    );
+    const updatedCurrentCombinedRows = getUpdatedCurrentCombinedRows(
+      updatedCombinedRows,
+      state.currentCombinedRows
+    );
+    scrollToEarliestCurrentRow(updatedCurrentCombinedRows, updatedCombinedRows);
+    dispatch({
+      type: "setAllCombinedRows",
+      value: updatedCombinedRows,
+    });
+    dispatch({
+      type: "setCurrentCombinedRows",
+      value: updatedCurrentCombinedRows,
+    });
+  };
+};
+
+const getUpdatedCurrentCombinedRows = (
+  allCombinedRows: Types.CombinedRow[],
+  currentCombinedRows: Types.CombinedRow[]
+): Types.CombinedRow[] => {
+  const childRows = Renderers.getFlattenedChildRows(currentCombinedRows);
+  const parentCombinedRows = childRows.map(findExistingParent(allCombinedRows));
+  const uniqueParents = parentCombinedRows.filter(firstInstanceOfCombinedRow);
+  return uniqueParents;
+};
+
+const findExistingParent = (combinedRows: Types.CombinedRow[]) => (
+  row: Types.Row
+): Types.CombinedRow => {
+  const existingParent = combinedRows.find(parentContainsGivenRow(row));
+  return !existingParent ? Config.defaultCombinedRow : existingParent;
+};
+
+const parentContainsGivenRow = (givenRow: Types.Row) => (
+  parentCombinedRow: Types.CombinedRow
+): Types.Row | undefined => {
+  return parentCombinedRow.rows.find((row) => row === givenRow);
+};
+
+const firstInstanceOfCombinedRow = (
+  givenRow: Types.CombinedRow,
+  index: number,
+  parentRows: Types.CombinedRow[]
+): boolean => {
+  return parentRows.indexOf(givenRow) === index;
+};
+
+const scrollToEarliestCurrentRow = (
```

Comment:
Related to "Got table scroll to jump correctly (most of the time?) to earliest persisted combinedRow in currentCombinedRows"